server/src/main/java/org/opensearch/search/aggregations/bucket/terms/MultiTermsAggregator.java

```diff
@@ -414,6 +372,74 @@ public void close() {
         }
     }
 
+    static class MultiValuesSourceCollectorImpl implements MultiTermsValuesSourceCollector {
+
+        private final List<InternalValuesSourceCollector> collectors;
+        private final BytesStreamOutput scratch;
+        private final BytesKeyedBucketOrds bucketOrds;
+        private final BucketsAggregator aggregator;
+        private final LeafBucketCollector sub;
+
+        private final boolean collectViaAggregator;
+
+        public MultiValuesSourceCollectorImpl(
+            List<InternalValuesSourceCollector> collectors,
+            BytesStreamOutput scratch,
+            BytesKeyedBucketOrds bucketOrds,
+            BucketsAggregator aggregator,
+            LeafBucketCollector sub
+        ) {
+            this.collectors = collectors;
+            this.scratch = scratch;
+            this.bucketOrds = bucketOrds;
+            this.aggregator = aggregator;
+            this.sub = sub;
+            this.collectViaAggregator = aggregator != null && sub != null;
+        }
+
+        @Override
+        public void apply(int doc, long owningBucketOrd) throws IOException {
+            List<List<TermValue<?>>> collectedValues = new ArrayList<>();
+            for (InternalValuesSourceCollector collector : collectors) {
+                collectedValues.add(collector.apply(doc));
+            }
+            scratch.seek(0);
+            scratch.writeVInt(collectors.size()); // number of fields per composite key
+            cartesianProductRecursive(collectedValues, 0, owningBucketOrd, doc);
+        }
+
+        /**
+         * Cartesian product using depth first search.
+         */
+        private void cartesianProductRecursive(List<List<TermValue<?>>> collectedValues, int index, long owningBucketOrd, int doc)
+            throws IOException {
+            if (collectedValues.size() == index) {
```

Comment:
nit: On code structuring, I think the calls to ` aggregator.collectExistingBucket` or ` aggregator.collectBucket` should be part of `apply()` and this method should only deal with getting the multi-valued bucketord. 


Also, should we rename this method to `multiValueCartesianProduct()` instead? or something similar.

Reply:
Moving this out would require returning a List of `bucketOrd` equal to the number of composite keys per document. 

I am bad at naming things, maybe `generateAndCollectCompositeKeys` ?